server/authenticator.rs

```diff
@@ -0,0 +1,40 @@
+use std::sync::Arc;
+
+use system::concepts::Credential;
+use tonic::{Request, Status};
+use user::user_manager::UserManager;
+
+#[derive(Debug)]
+pub struct Authenticator {
+    user_manager: Arc<UserManager>,
+}
+
+impl Authenticator {
+    pub(crate) fn new(user_manager: Arc<UserManager>) -> Self {
+        Self { user_manager }
+    }
+}
+
+impl Authenticator {
+    pub fn authenticate(&self, req: Request<()>) -> Result<Request<()>, Status> {
+        let metadata = req.metadata();
+        let username_metadata = metadata.get("username").map(|u| u.to_str());
+        let password_metadata = metadata.get("password").map(|u| u.to_str());
+        match (username_metadata, password_metadata) {
+            (Some(Ok(username)), Some(Ok(password))) => match self.user_manager.get(username) {
+                Some((_, Credential::PasswordType { password_hash })) => {
+                    if password_hash.matches(password) {
+                        Ok(req)
+                    } else {
+                        Err(Status::unauthenticated("invalid credential supplied"))
+                    }
+                }
+                None => Err(Status::unauthenticated("invalid credential supplied")),
+            },
+            _ => {
+                Ok(req)
+                // Err(Status::unauthenticated("credential must be supplied"))
```

Comment:
I think this should be the Err case right ("credential must be supplied") ? Otherwise, if you for isntance write your owner driver, which never sends the username/password, you just pass through here as if you're authenticated!

Reply:
Fixed in fdb755434435739522d94dfda7bb07647a304544.